### PR TITLE
CNV-32485: Use cluster-wise eviction strategy configuration

### PR DESCRIPTION
**Pull request by upalatucci**
## 📝 Description

EvictionStrategy on 4.14 can be configured cluster-wise using the Hyperconverge configuration

When VM does not specify an evictionStrategy, the vmi will be created with the configured strategy. 

This means that on other screens like the schedule tab, when we show 'No eviction strategy' we have to show the cluster strategy. 

Steps:

1. create a common hook to fetch cluster configuration
2. edit pending changes considering the cluster configuration
3. Create a common component to show the EvictionStrategy in all the schedule tab that use the cluster configuration (other than the template schedule tab with its specific component)
4. on checkbox uncheck, set the evictionStrategy to None instead of deleting it. Deleting the strategy means 'Use the cluster strategy'

## 🎥 Demo

**Before**

Screenshot from 2023-09-19 14-05-34
Screenshot from 2023-09-19 14-05-28

**After**

**reading from the cluster configuration**

Screenshot from 2023-09-19 14-04-48
Screenshot from 2023-09-19 14-04-42

### CODE REVIEW

File: src/utils/hooks/useHyperConvergeConfiguration.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  if (isEmpty(hyperConverged)) return null;
+  if (hyperConverged?.items) return hyperConverged?.items?.[0];
+  if (Array.isArray(hyperConverged)) return hyperConverged?.[0];
+  return hyperConverged;
+};
+
+const useHyperConvergeConfiguration = (): [
+  hyperConvergeConfig: HyperConverged,
+  loaded: boolean,
+  error: any,
```
Comment (metalice) [inline src/utils/hooks/useHyperConvergeConfiguration.ts:25]: try using Error

File: src/utils/hooks/useHyperConvergeConfiguration.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const getHyperConvergedObject = (hyperConverged): HyperConverged => {
+  if (isEmpty(hyperConverged)) return null;
+  if (hyperConverged?.items) return hyperConverged?.items?.[0];
+  if (Array.isArray(hyperConverged)) return hyperConverged?.[0];
+  return hyperConverged;
+};
+
+const useHyperConvergeConfiguration = (): [
```
Comment (metalice) [inline src/utils/hooks/useHyperConvergeConfiguration.ts:22]: please make type full function and return value

File: src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx [hunk 3]
```diff
@@ -26,17 +29,30 @@ const EvictionStrategyModal: React.FC<EvictionStrategyModalProps> = ({
   vmi,
 }) => {
   const { t } = useKubevirtTranslation();
-  const [checked, setChecked] = React.useState<boolean>(
-    !!vm?.spec?.template?.spec?.evictionStrategy,
-  );
+  const isLiveMigrate =
+    vm?.spec?.template?.spec?.evictionStrategy === EVICTION_STRATEGIES.LiveMigrate;
```
Comment (metalice) [inline src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx:33]: also a selector, please use a selector across all this pr

Comment (metalice) [reply src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx:33]: also u can move it into useeffect

File: src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx [hunk 4]
```diff
@@ -26,17 +29,30 @@ const EvictionStrategyModal: React.FC<EvictionStrategyModalProps> = ({
...
+
+  useEffect(() => {
+    if (vm?.spec?.template?.spec?.evictionStrategy || hyperLoadingError || !hyperLoaded) return;
+
+    if (hyperConverge?.spec?.evictionStrategy) {
+      setChecked(hyperConverge?.spec?.evictionStrategy === EVICTION_STRATEGIES.LiveMigrate);
+      return;
+    }
+
+    setChecked(EVICTION_STRATEGY_DEFAULT === EVICTION_STRATEGIES.LiveMigrate);
```
Comment (metalice) [inline src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx:46]: ?

File: src/views/clusteroverview/SettingsTab/ClusterTab/components/TemplatesProjectSection/TemplatesProjectSection.tsx [hunk 5]
```diff
@@ -56,33 +51,24 @@ const TemplatesProjectSection: FC = () => {
...
-      const currentNamespaceHCO = getCurrentTemplatesNamespaceFromHCO(hyperConvergeObject);
+    if (hyperConverge) {
+      const currentNamespaceHCO = getCurrentTemplatesNamespaceFromHCO(hyperConverge);
       !selectedProject && setSelectedProject(currentNamespaceHCO ?? OPENSHIFT);
     }
-  }, [hyperConvergeData, selectedProject]);
+  }, [hyperConverge, selectedProject]);
 
   const onSelect = (
     _event: React.ChangeEvent<Element> | React.MouseEvent<Element, MouseEvent>,
```
Comment (metalice) [inline src/views/clusteroverview/SettingsTab/ClusterTab/components/TemplatesProjectSection/TemplatesProjectSection.tsx:64]: if you already playing with this file can you please remove the React.* ?

File: src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx (hunk 4 above)
Comment (upalatucci) [reply src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx:46]: If the vm and the hyperconverge do not have a strategy, make it true. 
I can even use true directly

File: src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx [hunk 6]
```diff
@@ -26,17 +30,30 @@ const EvictionStrategyModal: React.FC<EvictionStrategyModalProps> = ({
   vmi,
 }) => {
   const { t } = useKubevirtTranslation();
-  const [checked, setChecked] = React.useState<boolean>(
-    !!vm?.spec?.template?.spec?.evictionStrategy,
-  );
 
-  const updatedVirtualMachine = React.useMemo(() => {
+  const [checked, setChecked] = useState<boolean>(true);
```
Comment (pcbailey) [inline src/utils/components/EvictionStrategy/EvictionStrategyModal.tsx:34]: nitpick: We typically use the `is<Whatever>` convention for boolean state. I would make this `isChecked` to remain consistent. There are other places where we use `checked`, but I think we should refactor those to match the convention.

File: src/views/templates/details/tabs/scheduling/components/EvictionStrategyModal.tsx [hunk 7]
```diff
@@ -15,21 +20,38 @@ type EvictionStrategyModalProps = {
...
   onClose,
   onSubmit,
   template,
 }) => {
   const { t } = useKubevirtTranslation();
-  const [checked, setChecked] = React.useState<boolean>(!!getEvictionStrategy(template));
+  const templateEvictionStrategy = getEvictionStrategy(template);
 
-  const updatedTemplate = React.useMemo(() => {
+  const [checked, setChecked] = useState<boolean>(
```
Comment (pcbailey) [inline src/views/templates/details/tabs/scheduling/components/EvictionStrategyModal.tsx:32]: Same as above.

File: src/views/templates/details/tabs/scheduling/components/EvictionStrategyModal.tsx [hunk 8]
```diff
@@ -15,21 +20,38 @@ type EvictionStrategyModalProps = {
...
+  }, [hyperConverge, hyperLoaded, hyperLoadingError, templateEvictionStrategy]);
+
+  const updatedTemplate = useMemo(() => {
     return produce<V1Template>(template, (templateDraft: V1Template) => {
       const draftVM = getTemplateVirtualMachineObject(templateDraft);
       checked
-        ? (draftVM.spec.template.spec.evictionStrategy = 'LiveMigrate')
-        : delete draftVM.spec.template.spec.evictionStrategy;
+        ? (draftVM.spec.template.spec.evictionStrategy = EVICTION_STRATEGIES.LiveMigrate)
+        : (draftVM.spec.template.spec.evictionStrategy = 'None');
```
Comment (pcbailey) [inline src/views/templates/details/tabs/scheduling/components/EvictionStrategyModal.tsx:54]: Consider making this a common constant.

**Review by pcbailey — COMMENTED** [review]
It's looking great! I just had a few small suggestions.

**Pull request merged**